services/gateway/src/ops_batcher.rs

```
@@ -262,18 +283,156 @@ impl OpsBatcherRunner {
+                                                target_hex: &str,
+                                                out: &mut Vec<&'a serde_json::Value>,
+                                            ) {
+                                                if let Some(obj) = v.as_object() {
+                                                    if let Some(to) =
+                                                        obj.get("to").and_then(|t| t.as_str())
+                                                    {
+                                                        if to.eq_ignore_ascii_case(target_hex) {
+                                                            out.push(v);
+                                                        }
+                                                    }
+                                                    if let Some(calls) =
+                                                        obj.get("calls").and_then(|c| c.as_array())
+                                                    {
+                                                        for c in calls {
+                                                            collect_calls(c, target_hex, out);
+                                                        }
+                                                    }
+                                                }
+                                            }
+
+                                            let target_hex = format!("0x{:x}", registry);
+                                            let mut subcalls: Vec<&serde_json::Value> = Vec::new();
+                                            let root = trace.get("result").unwrap_or(&trace);
+                                            collect_calls(root, &target_hex, &mut subcalls);
+
+                                            // If counts mismatch, fall back to finalize-all to avoid mislabeling.
+                                            if subcalls.len() != ids_for_receipt.len() {
```

This should trigger an alert to investigate. This should never happen

Lets throw an error and put in error state.